lib/Pipes/PipelineManager.cpp

```diff
@@ -157,11 +164,124 @@ checkComponentsVersion(const revng::DirectoryPath &ExecutionDirectory,
+                                  { "model",
+                                    "migrate",
+                                    "--input",
+                                    InputTemporaryFile.path().str(),
+                                    "--output",
+                                    OutputTemporaryFile.path().str() });
+
+    if (ReturnCode != 0)
+      return revng::createError("Error encountered while running migrations, "
+                                "process returned: %d",
+                                ReturnCode);
+  }
+
+  revng::FilePath
+    OutputFile = revng::FilePath::fromLocalStorage(OutputTemporaryFile.path());
+  if (auto Error = OutputFile.copyTo(ModelFile); !!Error)
+    return Error;
+
+  return llvm::Error::success();
+}
+
+// Conveninece function for setUpPipeline, creates a backup for the model file
+// and runs migrations on it. On success returns the backup path, on failure
+// returns the error and restores the model file from backup. If something goes
+// horribly wrong and it fails to restore, it aborts.
+static llvm::Expected<revng::FilePath>
+migrateModel(const revng::DirectoryPath &ExecutionDirectory) {
+  auto ModelFile = ExecutionDirectory.getDirectory("context")
+                     .getFile(revng::ModelGlobalName);
+
+  if (not ModelFile.exists()) {
+    // Model file does not exist, so no migrations will actually run (empty
+    // model is assumed to be in the latest version). So for the sake of better
+    // error reporting, process this case explicitly.
+    return revng::createError("Migrations will not be run since the model file "
+                              "does not exist");
+  }
+
+  llvm::Expected<revng::FilePath>
+    ModelBackupFile = createModelBackup(ModelFile);
+
+  if (not ModelBackupFile)
+    return ModelBackupFile.takeError();
+
+  if (auto MigrationError = migrateModelFile(ModelFile); !!MigrationError) {
+    // Migration failed, rollback to the previous version of the model.
+    if (auto CopyError = ModelBackupFile->copyTo(ModelFile); !!CopyError) {
+      // BIG PROBLEM: model likely changed and retrieval from backup failed.
+      revng_abort("Model migration failed and the model could not be rolled "
+                  "back to the original version, this requires some manual "
+                  "intervention. The model backups can be found alongside the "
+                  "model file.");
+    }
+
+    return MigrationError;
+  }
+
+  return ModelBackupFile;
+}
+
+static llvm::Error
+restoreModelFromBackup(const revng::DirectoryPath &ExecutionDirectory,
+                       const revng::FilePath &BackupFilePath) {
+  auto ModelFile = ExecutionDirectory.getDirectory("context")
+                     .getFile(revng::ModelGlobalName);
+
+  if (auto Error = BackupFilePath.copyTo(ModelFile))
+    return Error;
+
+  return llvm::Error::success();
+}
+
+static llvm::Expected<pipeline::Runner>
 setUpPipeline(pipeline::Context &PipelineContext,
```

Comment:
`setupPipeline` is accepting basically all the properties of `PipelineManager`, make it a private method of it.

Reply:
Thanks

> On a similar note, it might also make sense to refactor createFromMemory into private member functions.

What do you mean? The constructor of `PipelineManager` is private, the way users are expected to instantiate it is via `create` or `createFromMemory`

Reply:
I see. I had the impression that `create` is the intended user-facing function and `createFromMemory` are utilities for it. If `createFromMemory` is also user-facing, then my suggestion does not apply.

Marking as resolved.